common/src/main/java/org/apache/celeborn/common/network/server/BufferStreamManager.java

```diff
@@ -313,74 +321,118 @@ public synchronized void setupDataPartitionReader(
               streamId,
               channel,
               () -> recycleStream(streamId));
-      // allocate resources when the first reader is registered
-      boolean allocateResources = readers.isEmpty();
-      readers.add(dataPartitionReader);
-      streamReaders.put(streamId, dataPartitionReader);
+      readers.put(streamId, dataPartitionReader);
 
-      // create initial buffers for read
-      if (allocateResources && buffers == null) {
+      // allocate resources when the first reader is registered
+      if (firstReaderRegister.compareAndSet(true, false)) {
```

Comment:
firstReaderRegister needn't to be an atomicBoolean as it only access by one thread at a time. And the name would be better rename to bufferInitialized or bufferXXX for clarify the purpose